### PR TITLE
Extension - Add/Remove Chains (Smoldot v.0.3.x)

### LINKED ISSUE #427: Extension for smoldot v.0.3.X - Add/Remove Chains
When an app is connected in Smoldot through the extension, it should call Smoldot with the chain specs, no matter if the chain is a relay chain or not. Smoldot will understand and keep only 1 chain connected.
During connection the Extension should keep an Id on the app for each Chain that is connected to.
Upon disconnection the Extension should call again Smoldot with `remove()` function, for the specific chain that app was connected to.

### PR DESCRIPTION
**Pull request by wirednkod**
Fixes #427

### CODE REVIEW

**Review by raoulmillais — CHANGES_REQUESTED** [review]
In general I think this is a really good start :+1: The biggest issue I can see so far is that I really don't think we should be sending the `spec` message using the `method` field of the RPC message. We don't own the JSON RPC protocol that is the domain of smoldot.  We have our own `type` field in the `MessageToManager` which is designed to ask the background - you should extend that instead and use `provider.send()` to send it from the `ExtensionProvider`.

Aside: Our logging strategy is going a bit all over the place in the extension, but I don't think you need to fix that in this PR: we need aticket for that though.

**Comment by wirednkod** [comment]
> Aside: Our logging strategy is going a bit all over the place in the extension, but I don't think you need to fix that in this PR: but we need a ticket for it though.

Add new issue #429

File: packages/connect-extension-protocol/src/index.ts [hunk 1]
```diff
@@ -68,7 +68,7 @@ export interface ProviderMessageData {
   /** The name of the blockchain network the app is talking to **/
   chainName: string;
   /** What action the `ExtensionMessageRouter` should take **/
-  action: 'forward' | 'connect' | 'disconnect';
+  action: 'forward' | 'connect' | 'disconnect' | 'spec';
```
Comment (raoulmillais) [inline packages/connect-extension-protocol/src/index.ts:71]: I don't think we want a new `action` at this layer.  We want a message from the `ExtensionProvider` to ask the `ExtensionMessage` to **`forward`** a `'spec'` message.

File: packages/connect/src/ExtensionProvider/ExtensionProvider.ts [hunk 2]
```diff
@@ -216,8 +216,19 @@ export class ExtensionProvider implements ProviderInterface {
 
     // Once connect is sent - send rpc to extension that will contain the chainSpecs
     // for the extension to call addChain on smoldot
-    this.send('spec', [this.#chainSpecs]).catch(console.error);
-    
+    const someMsg: ProviderMessageData = {
+      appName: this.#appName,
+      chainName: this.#chainName,
+      action: 'spec',
```
Comment (raoulmillais) [inline packages/connect/src/ExtensionProvider/ExtensionProvider.ts:222]: See above

File: projects/extension/src/content/ExtensionMessageRouter.ts [hunk 3]
```diff
@@ -112,6 +124,12 @@ export class ExtensionMessageRouter {
       return this.#establishNewConnection(message);
     }
 
+    if (data.action === 'spec') {
+      console.log('data action: ', data.action);
+      console.log('message', message.data)
```
Comment (raoulmillais) [inline projects/extension/src/content/ExtensionMessageRouter.ts:129]: Change these to use the `debug` helper (or remove them)

**Review by raoulmillais — CHANGES_REQUESTED** [review]

File: projects/extension/src/content/ExtensionMessageRouter.ts (hunk 3 above)
Comment (wirednkod) [reply projects/extension/src/content/ExtensionMessageRouter.ts:129]: Sorry about these stray logs - will be removed

File: packages/connect-extension-protocol/src/index.ts (hunk 1 above)
Comment (wirednkod) [reply packages/connect-extension-protocol/src/index.ts:71]: The idea behind this was to have complete different flows for spec and other json-rpc forwards for future usage - but if it makes sense to have flow until split this on the message level - I will proceed with that

File: projects/extension/src/background/AppMediator.ts [hunk 4]
```diff
@@ -261,14 +261,14 @@ export class AppMediator extends (EventEmitter as { new(): StateEmitter }) {
     }
     const chainName = this.#chainName as string;
 
-    if (parsed.method === 'spec' && chainName) {
+    if (msg.type === 'spec' && chainName) {
       // When params[0] (chainSpecs in the current case is empty) then this is a
```
Comment (raoulmillais) [inline projects/extension/src/background/AppMediator.ts:265]: This comment is no longer valid

**Review by raoulmillais — APPROVED** [review]
Just one little stray comment to remove, and this is good to merge :+1:

File: projects/extension/src/background/AppMediator.ts [hunk 5]
```diff
@@ -14,6 +13,18 @@ import {
   StateEmitter,
   SubscriptionMapping
 } from './types';
+import { SmoldotChain } from 'smoldot';
+import westend from '../../public/assets/westend.json';
+import kusama from '../../public/assets/kusama.json';
+import polkadot from '../../public/assets/polkadot.json';
+
+type RelayType = Record<string, string>
```
Comment (raoulmillais) [inline projects/extension/src/background/AppMediator.ts:21]: Not a blocker .. but this should be a `Map<string, string>`

Comment (wirednkod) [reply projects/extension/src/background/AppMediator.ts:21]: This does not work:
```
Type '{ polkadot: string; kusama: string; westend: string; }' is not assignable to type 'RelayType'.
  Object literal may only specify known properties, and '"polkadot"' does not exist in type 'Map<string, string>'.ts(2322)
```

**Pull request merged**